stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -390,36 +415,54 @@ impl<
     /// If we process a PoX anchor block, then return its block hash.  This unblocks processing the
-    /// next reward cycle's burnchain blocks.
+    /// next reward cycle's burnchain blocks.  Subsequent calls to this function will terminate
+    /// with Some(pox-anchor-block-hash) until the reward cycle info is processed in the sortition
+    /// DB.
     pub fn handle_new_nakamoto_stacks_block(&mut self) -> Result<Option<BlockHeaderHash>, Error> {
         let canonical_sortition_tip = self.canonical_sortition_tip.clone().expect(
             "FAIL: processing a new Stacks block, but don't have a canonical sortition tip",
         );
 
         loop {
             // process at most one block per loop pass
-            let sortdb_handle = self
+            let mut sortdb_handle = self
                 .sortition_db
                 .tx_handle_begin(&canonical_sortition_tip)?;
 
-            let mut processed_blocks = NakamotoChainState::process_nakamoto_blocks(
+            let mut processed_block_receipt = match NakamotoChainState::process_next_nakamoto_block(
                 &mut self.chain_state_db,
-                sortdb_handle,
-                1,
+                &mut sortdb_handle,
                 self.dispatcher,
-            )?;
+            ) {
+                Ok(receipt_opt) => receipt_opt,
+                Err(ChainstateError::InvalidStacksBlock(msg)) => {
+                    warn!("Encountered invalid block: {}", &msg);
+
+                    // try again
+                    self.notifier.notify_stacks_block_processed();
+                    increment_stx_blocks_processed_counter();
+                    continue;
+                }
+                Err(ChainstateError::NetError(NetError::DeserializeError(msg))) => {
+                    // happens if we load a zero-sized block (i.e. an invalid block)
+                    warn!("Encountered invalid block (codec error): {}", &msg);
+                    
+                    // try again
+                    self.notifier.notify_stacks_block_processed();
+                    increment_stx_blocks_processed_counter();
+                    continue;
+                }
+                Err(e) => {
+                    // something else happened
+                    return Err(e.into());
+                }
+            };
 
-            if processed_blocks.len() == 0 {
+            sortdb_handle.commit()?;
+
+            let Some(block_receipt) = processed_block_receipt.take() else {
                 // out of blocks
                 break;
-            }
-
-            let Some(Some(block_receipt)) = processed_blocks.pop() else {
-                // this block was invalid
-                debug!("Bump blocks processed (invalid)");
-                self.notifier.notify_stacks_block_processed();
-                increment_stx_blocks_processed_counter();
-                continue;
             };
 
             // only bump the coordinator's state if the processed block
```

Comment:
I think this check needs to be dropped -- the block_hash may not show up in the sortition-fork yet.